aiida_vasp/workchains/vasp.py

```diff
@@ -375,29 +418,105 @@ def _handle_unfinished_calculation(self, node):
+                    self._setup_restart(node)
+                    self.ctx.inputs.parameters.update(incar)
+                    return ProcessHandlerReport(do_break=True)
+            # Change to ALGO if options have been exhausted
+            incar['algo'] = 'all'
+            self.ctx.inputs.parameters.update(incar)
+            self._setup_restart(node)
+            return ProcessHandlerReport(do_break=True)
+
+        return ProcessHandlerReport(do_break=True,
+                                    exit_code=self.exit_codes.ERROR_MANUAL_INTERVENTION_NEEDED.format(
+                                        message='Cannot apply fix for reaching eletronic convergence.'))
+
+    @process_handler(priority=50, exit_codes=[VaspCalculation.exit_codes.ERROR_IONIC_NOT_CONVERGED])
+    def _handle_ionic_converge_problem(self, node):
+        """Handle ionic convergence problem"""
+        if 'structure' not in node.outputs.structure:
+            self.report('Performing a geometry optimisation but the output structure is not found.')
+            return ProcessHandlerReport(do_break=True,
+                                        exit_code=self.exit_codes.ERROR_MANUAL_INTERVENTION_NEEDED.format(
+                                            message='No output structure for restarting ionic relaxation.'))  #pylint: disable=no-member
+        # The simplest solution - resubmit the calculation again
+        self.report('Continuing geometry optimisation using the last geometry.')
+        self.ctx.inputs.structure = node.outputs.structure
+        self._setup_restart(node)
+        return ProcessHandlerReport(do_break=True)
+
+    @process_handler(priority=90, exit_codes=[VaspCalculation.exit_codes.ERROR_VASP_CRITICAL_ERROR])
+    def _handle_vasp_critical(self, node):
+        """Handle critical errors"""
+        notification = node.outputs.misc['notifications']
+        self.report('Critical notifications detected: {} - aborting'.format(notification))
+        return ProcessHandlerReport(do_break=True,
+                                    exit_code=self.exit_codes.ERROR_MANUAL_INTERVENTION_NEEDED.format(message=', '.join(notification)))
+
+    def _setup_restart(self, node):
         """
-        Attach WAVECAR for the new calculation if there is one.
+        Attach WAVECAR for the new calculation, if there is one.
         """
-        # Check the exists of the WAVECAR
-        has_wavecar = False
-        for file in node.outputs.remote_folder.listdir_withattributes():
-            if file['name'] == 'WAVECAR' and file['attributes'].st_size > 1024 * 16:
-                has_wavecar = True
-        if has_wavecar:
-            self.report('Found WAVECAR in the last calcualtion folder, using the outputs for restart.')
+        if self.ctx.last_calc_remote_files is None:
+            self.ctx.last_calc_remote_files = list_valid_files_in_remote(node.outputs.remote_folder)
+        if 'WAVECAR' in self.ctx.last_calc_remote_files:
```

Comment:
Again, this goes back to the handler with the mixing tuning etc. I think we really need to lift this out of the handlers as we need to be able to provide the provenance as clearly as possible to the users. Also, we need to be able to not start from an old `WAVECAR`, even though it exists. Not sure how this logic works if we just adjust the `ISTART` to ignore `WAVECAR` in a higher workchain? Would it enable restarts?

Reply:
I think we really need to lift this out of the handlers as we need to be able to provide the provenance as clearly as possible to the users.

I don't think so, with the handlers the provenance is still there, the messages are emitted and the input parameters are changed, and possibly the remote folder linked to. Otherwise, it is just more complications, and no implementations.
There is also an option to store the list of activated handlers, but I already think it is going too far...

Do we not want to restart when there is no WAVECAR? Also, we would as part of the electronic convergence sometimes need to not start from the existing WAVECAR.

I see. I will add an option to disable using the WAVECAR. I think most of the time it should be used to accelerate the convergence (say restarting an electronically unconverged calculation)?

Reply:
> I don't think so, with the handlers the provenance is still there, the messages are emitted and the input parameters are changed, and possibly the remote folder linked to. Otherwise, it is just more complications, and no implementations.
There is also an option to store the list of activated handlers, but I already think it is going too far...

Yes, maybe. But this would in the end be similar to the parameter popup discussions with respect to `ParameterMassager` etc. Let us leave it and I will consider to lift it out if we find use for that down the line.

> I see. I will add an option to disable using the WAVECAR. I think most of the time it should be used to accelerate the convergence (say restarting an electronically unconverged calculation)?

Thanks a lot for adding this.